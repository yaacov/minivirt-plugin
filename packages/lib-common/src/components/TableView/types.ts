import { Field } from '../types';

export interface Column {
  id: string;
  toLabel(t: (key: string) => string): string;
  sortable?: boolean;
}

export interface RowProps<T> {
  columns: Field[];
  entity: T;
}
