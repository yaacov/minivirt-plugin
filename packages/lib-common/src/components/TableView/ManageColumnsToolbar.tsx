import React, { useState } from 'react';
import { useTranslation } from '../../utils/i18n';

import {
  Button,
  DataList,
  DataListCell,
  DataListCheck,
  DataListControl,
  DataListDragButton,
  DataListItem,
  DataListItemCells,
  DataListItemRow,
  DragDrop,
  Draggable,
  Droppable,
  Modal,
  Text,
  TextContent,
  TextVariants,
  ToolbarItem,
  Tooltip,
} from '@patternfly/react-core';
import { ColumnsIcon } from '@patternfly/react-icons';

import { Field } from '../types';

export interface ManageColumnsToolbarProps {
  /** Read only. State maintained by parent component. */
  columns: Field[];
  /** Read only. The defaults used for initialization.*/
  defaultColumns: Field[];
  /** Setter to modify state in the parent.*/
  setColumns(columns: Field[]): void;
}

/**
 * Toggles a modal dialog for managing columns visibility and order.
 */
export const ManageColumnsToolbar = ({
  columns,
  setColumns,
  defaultColumns,
}: ManageColumnsToolbarProps) => {
  const { t } = useTranslation();
  const [isOpen, setIsOpen] = useState(false);
  return (
    <ToolbarItem>
      <Tooltip content={t('Manage Columns')}>
        <Button variant="plain" onClick={() => setIsOpen(true)} aria-label={t('Manage Columns')}>
          <ColumnsIcon />
        </Button>
      </Tooltip>
      <ManageColumns
        showModal={isOpen}
        onClose={() => setIsOpen(false)}
        description={t('Selected columns will be displayed in the table.')}
        columns={columns}
        onChange={setColumns}
        defaultColumns={defaultColumns}
      />
    </ToolbarItem>
  );
};

interface ManagedColumnsProps {
  showModal: boolean;
  description: string;
  onClose(): void;
  /** Setter to modify state in the parent.*/
  onChange(colums: Field[]): void;
  /** Read only. State maintained by parent component. */
  columns: Field[];
  /** Read only. The defaults used for initialization.*/
  defaultColumns: Field[];
}
/**
 * Modal dialog for managing columns.
 * Supported features:
 * 1. toggle column visibility (disabled for identity columns that should always be displayed to uniquely identify a row)
 * 2. re-order the columns using drag and drop
 */
const ManageColumns = ({
  showModal,
  description,
  onClose,
  onChange,
  columns,
  defaultColumns,
}: ManagedColumnsProps) => {
  const { t } = useTranslation();
  const [editedColumns, setEditedColumns] = useState(columns);
  const restoreDefaults = () => setEditedColumns([...defaultColumns]);
  const onDrop = (source: { index: number }, dest: { index: number }) => {
    const draggedItem = editedColumns[source?.index];
    const itemCurrentlyAtDestination = editedColumns[dest?.index];
    if (!draggedItem || !itemCurrentlyAtDestination) {
      return false;
    }
    const base = editedColumns.filter(({ id }) => id !== draggedItem?.id);
    setEditedColumns([
      ...base.slice(0, dest.index),
      draggedItem,
      ...base.slice(dest.index, base.length),
    ]);
    return true;
  };
  const onSelect = (updatedId: string, updatedValue: boolean): void => {
    setEditedColumns(
      editedColumns.map(({ id, isVisible, ...rest }) => ({
        id,
        ...rest,
        isVisible: id === updatedId ? updatedValue : isVisible,
      })),
    );
  };
  const onSave = () => {
    onChange(editedColumns);
    onClose();
  };

  return (
    <Modal
      title={t('Manage Columns')}
      isOpen={showModal}
      variant="small"
      description={
        <TextContent>
          <Text component={TextVariants.p}>{description}</Text>
        </TextContent>
      }
      onClose={onClose}
      actions={[
        <Button
          key="save"
          variant="primary"
          isDisabled={columns === editedColumns}
          onClick={onSave}
        >
          {t('Save')}
        </Button>,
        <Button key="cancel" variant="secondary" onClick={onClose}>
          {t('Cancel')}
        </Button>,
        <Button key="restore" variant="link" onClick={restoreDefaults}>
          {t('Restore default colums')}
        </Button>,
      ]}
    >
      <DragDrop onDrop={onDrop}>
        <Droppable hasNoWrapper>
          <DataList
            aria-label={t('Table column management')}
            id="table-column-management"
            isCompact
          >
            {editedColumns.map(({ id, isVisible, isIdentity, toLabel }) => (
              <Draggable key={id} hasNoWrapper>
                <DataListItem aria-labelledby={`draggable-${id}`} ref={React.createRef()}>
                  <DataListItemRow>
                    <DataListControl>
                      <DataListDragButton
                        aria-label={t('Reorder')}
                        aria-labelledby={`draggable-${id}`}
                      />
                      <DataListCheck
                        aria-labelledby={`draggable-${id}`}
                        name={id}
                        checked={isVisible}
                        isDisabled={isIdentity}
                        onChange={(value) => onSelect(id, value)}
                        otherControls
                      />
                    </DataListControl>
                    <DataListItemCells
                      dataListCells={[
                        <DataListCell key={id}>
                          <span id={`draggable-${id}`}>{toLabel(t)}</span>
                        </DataListCell>,
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>
              </Draggable>
            ))}
          </DataList>
        </Droppable>
      </DragDrop>
    </Modal>
  );
};
