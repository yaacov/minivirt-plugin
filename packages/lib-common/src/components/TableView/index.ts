export * from './ManageColumnsToolbar';
export * from './TableView';
export * from './types';
