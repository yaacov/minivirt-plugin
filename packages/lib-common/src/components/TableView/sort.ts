import { useMemo, useState } from 'react';
import { useTranslation } from '../../utils/i18n';

import { ThSortType } from '@patternfly/react-table/dist/esm/components/Table/base';

import { Field, SortType } from '../types';

import { Column } from './types';

/**
 * Uses native string localCompare method with numeric option enabled.
 *
 * @param locale to be used by string comparator
 */
 export const localeCompare = (a: string, b: string, locale: string): number =>
 a.localeCompare(b, locale, { numeric: true });

/**
 * Compares all types by converting them to string.
 * Nullish entities are converted to empty string.
 * @see localeCompare
 * @param locale to be used by string comparator
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const universalComparator = (a: any, b: any, locale: string) =>
  localeCompare(String(a ?? ''), String(b ?? ''), locale);

/**
 * Creates a comparator function based on provided current sort definition.
 * If there is no sort defined then comparator considers all entities equal.
 *
 * @see universalComparator
 * @param currentSort
 * @param locale defaults to "en"
 * @param fieldComparator (optional) custom field comparator. Defaults to universal string based comparator.
 * @returns comparator function
 */
export function compareWith(
  currentSort: SortType,
  locale: string,
  fieldComparator: (a, b, locale: string) => number,
): (a, b) => number {
  return (a, b) => {
    if (!currentSort?.id) {
      return 0;
    }
    const comparator = fieldComparator ?? universalComparator;
    const compareValue = comparator(a?.[currentSort.id], b?.[currentSort.id], locale ?? 'en');
    return currentSort.isAsc ? compareValue : -compareValue;
  };
}

/**
 * Hook for maintaining sort state. Supported features:
 * 1) by default sort by the first identity column or by the first column available if there is no identity column
 * 2) build comparator based on the current active sort definition
 *
 * @param fields (read only) field metadata
 * @returns [activeSort, setActiveSort, comparator]
 */
export const useSort = (
  fields: Field[],
): [SortType, (sort: SortType) => void, (a, b) => number] => {
  const { i18n } = useTranslation();

  // by default sort by the first identity column (if any)
  const [firstField] = [...fields].sort(
    (a, b) => Number(Boolean(b.isIdentity)) - Number(Boolean(a.isIdentity)),
  );

  const [activeSort, setActiveSort] = useState<SortType>({
    isAsc: false,
    id: firstField?.id,
    toLabel: firstField?.toLabel,
  });

  const comparator = useMemo(
    () =>
      compareWith(
        activeSort,
        i18n.resolvedLanguage,
        fields.find((field) => field.id === activeSort.id)?.comparator,
      ),
    [fields, activeSort],
  );

  return [activeSort, setActiveSort, comparator];
};

/**
 * Builds table specific sort definition based on provided active sort definition.
 * @see ThSortType
 */
export const buildSort = ({
  columnIndex,
  columns,
  activeSort,
  setActiveSort,
}: {
  columnIndex: number;
  columns: Column[];
  activeSort: SortType;
  setActiveSort: (sort: SortType) => void;
}): ThSortType => ({
  sortBy: {
    index:
      columns.find(({ id }) => id === activeSort.id) &&
      columns.findIndex(({ id }) => id === activeSort.id),
    direction: activeSort.isAsc ? 'asc' : 'desc',
  },
  onSort: (_event, index, direction) => {
    columns[index]?.id &&
      setActiveSort({
        isAsc: direction === 'asc',
        ...columns[index],
      });
  },
  columnIndex,
});
