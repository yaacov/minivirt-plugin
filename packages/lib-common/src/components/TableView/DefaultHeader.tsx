import React from 'react';
import { useTranslation } from '../../utils/i18n';

import { Th } from '@patternfly/react-table';

import { buildSort } from './sort';
import { TableViewHeaderProps } from './types';

export const DefaultHeader = ({
  visibleColumns,
  setActiveSort,
  activeSort,
}: TableViewHeaderProps) => {
  const { t } = useTranslation();
  return (
    <>
      {visibleColumns.map(({ id, toLabel, sortable }, columnIndex) => (
        <Th
          key={id}
          sort={
            sortable &&
            buildSort({
              activeSort,
              columnIndex,
              columns: visibleColumns,
              setActiveSort,
            })
          }
        >
          {toLabel(t)}
        </Th>
      ))}
    </>
  );
};
