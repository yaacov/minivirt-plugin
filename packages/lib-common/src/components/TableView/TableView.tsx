import React, { ReactNode } from 'react';
import { useTranslation } from '../../utils/i18n';

import { Bullseye } from '@patternfly/react-core';
import { TableComposable, Tbody, Td, Th, Thead, Tr } from '@patternfly/react-table';

import { Field, SortType } from '../types';

import { buildSort } from './sort';
import { RowProps } from './types';

/**
 * Displays provided list of entities as table. Supported features:
 * 1) sorting via arrow buttons in the header
 * 2) stable row keys based on entity[uidFieldId]
 * 3) (if present) display nodes passed via children prop instead of entities (extension point to handle empty state end related corner cases)
 *
 * @see useSort
 */
export function TableView<T>({
  uidFieldId = 'uid',
  visibleColumns,
  entities,
  'aria-label': ariaLabel,
  Row,
  children,
  activeSort,
  setActiveSort,
}: TableViewProps<T>) {
  const { t } = useTranslation();
  const hasChildren = children.filter(Boolean).length > 0;
  const columnSignature = visibleColumns.map(({ id }) => id).join();

  return (
    <TableComposable aria-label={ariaLabel} variant="compact" isStickyHeader>
      <Thead>
        <Tr>
          {visibleColumns.map(({ id, toLabel, sortable }, columnIndex) => (
            <Th
              key={id}
              sort={
                sortable &&
                buildSort({
                  activeSort,
                  columnIndex,
                  columns: visibleColumns,
                  setActiveSort,
                })
              }
            >
              {toLabel(t)}
            </Th>
          ))}
        </Tr>
      </Thead>
      <Tbody>
        {hasChildren && (
          <Tr>
            <Td colSpan={visibleColumns.length || 1}>
              <Bullseye>{children}</Bullseye>
            </Td>
          </Tr>
        )}
        {!hasChildren &&
          entities.map((entity, index) => (
            <Row
              key={`${columnSignature}_${entity?.[uidFieldId] ?? index}`}
              entity={entity}
              columns={visibleColumns}
            />
          ))}
      </Tbody>
    </TableComposable>
  );
}

interface TableViewProps<T> {
  visibleColumns: Field[];
  entities: T[];
  'aria-label': string;
  /**
   * entity[uidFieldId] is used to uniquely identify a row. Defaults to UID column.
   */
  uidFieldId?: string;
  /**
   * Maps entities to table rows.
   */
  Row(props: RowProps<T>): JSX.Element;
  /**
   * Nodes to be displayed instead of the entities.
   * Extension point to handle empty state and related cases.
   */
  children?: ReactNode[];
  activeSort: SortType;
  setActiveSort: (sort: SortType) => void;
}
