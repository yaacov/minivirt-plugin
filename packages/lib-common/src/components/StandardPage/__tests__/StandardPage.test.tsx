import * as React from 'react';
import { RowProps } from '../../TableView';

import { Td, Tr } from '@patternfly/react-table';
import { cleanup, render } from '@testing-library/react';

import { StandardPage } from '..';

afterEach(cleanup);

function SimpleRow<T>({ columns, entity }: RowProps<T>) {
  return (
    <Tr>
      {columns.map(({ id, toLabel }) => (
        <Td key={id} dataLabel={toLabel((s) => s)}>
          {String(entity[id] ?? '')}
        </Td>
      ))}
    </Tr>
  );
}

test('empty result set returned, no filters ', async () => {
  interface Named {
    name: string;
  }
  const dataSource: [Named[], boolean, boolean] = [[], true, false];
  const { asFragment } = render(
    <StandardPage<Named>
      RowMapper={SimpleRow}
      dataSource={dataSource}
      fieldsMetadata={[
        {
          id: 'name',
          toLabel: (t) => t('Name'),
        },
      ]}
      namespace={undefined}
      title="Simple"
    ></StandardPage>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('single entry returned, both filters ', async () => {
  interface Simple {
    name: string;
    namespace: string;
  }
  const dataSource: [Simple[], boolean, boolean] = [
    [{ name: 'foo_name', namespace: 'bar_namespace' }],
    true,
    false,
  ];
  const { asFragment, queryByText } = render(
    <StandardPage<Simple>
      RowMapper={SimpleRow}
      dataSource={dataSource}
      fieldsMetadata={[
        {
          id: 'name',
          toLabel: (t) => t('Name'),
          isIdentity: true,
          isVisible: true,
          filter: {
            primary: true,
            type: 'freetext',
            toPlaceholderLabel: (t) => t('Filter by name'),
          },
        },
        {
          id: 'namespace',
          toLabel: (t) => t('Namespace'),
          isIdentity: true,
          isVisible: true,
          filter: {
            type: 'freetext',
            toPlaceholderLabel: (t) => t('Filter by namespace'),
          },
        },
      ]}
      namespace={'some_namespace'}
      title="Simple"
    ></StandardPage>,
  );
  expect(queryByText('foo_name')).toBeTruthy();
  // namespace column should be hidden
  expect(queryByText('bar_namespace')).toBeFalsy();
  expect(asFragment()).toMatchSnapshot();
});
