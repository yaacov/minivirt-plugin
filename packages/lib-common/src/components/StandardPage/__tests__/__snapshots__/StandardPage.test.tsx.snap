// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty result set returned, no filters  1`] = `
<DocumentFragment>
  <section
    class="pf-c-page__main-section pf-m-light"
  >
    <div
      class="pf-l-level"
    >
      <div>
        <h1
          class="pf-c-title pf-m-2xl"
          data-ouia-component-id="OUIA-Generated-Title-1"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          Simple
        </h1>
      </div>
    </div>
  </section>
  <section
    class="pf-c-page__main-section"
  >
    <div
      class="pf-c-toolbar"
      data-ouia-component-id="OUIA-Generated-Toolbar-1"
      data-ouia-component-type="PF4/Toolbar"
      data-ouia-safe="true"
      id="pf-random-id-0"
    >
      <div
        class="pf-c-toolbar__content"
      >
        <div
          class="pf-c-toolbar__content-section"
        >
          <div
            class="pf-c-toolbar__group pf-m-toggle-group pf-m-show-on-xl"
          >
            <div
              class="pf-c-toolbar__toggle"
            >
              <button
                aria-controls="pf-random-id-0-expandable-content-0"
                aria-disabled="false"
                aria-haspopup="false"
                aria-label="Show Filters"
                class="pf-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-1"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="pf-c-toolbar__group pf-m-filter-group"
            />
            <div
              class="pf-c-toolbar__group pf-m-filter-group"
            >
              <div
                class="pf-c-toolbar__item"
              >
                <div
                  class="pf-c-select"
                  data-ouia-component-id="OUIA-Generated-Select-single-1"
                  data-ouia-component-type="PF4/Select"
                  data-ouia-safe="true"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-label="Options menu"
                    aria-labelledby=" pf-select-toggle-id-0"
                    class="pf-c-select__toggle"
                    id="pf-select-toggle-id-0"
                    type="button"
                  >
                    <div
                      class="pf-c-select__toggle-wrapper"
                    >
                      <span
                        class="pf-c-select__toggle-text"
                      />
                    </div>
                    <span
                      class="pf-c-select__toggle-arrow"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="pf-c-toolbar__item"
            >
              <button
                aria-disabled="false"
                aria-label="Manage Columns"
                class="pf-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-2"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64V160h160v256zm224 0H288V160h160v256z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="pf-c-toolbar__expandable-content"
          id="pf-random-id-0-expandable-content-0"
        >
          <div
            class="pf-c-toolbar__group"
          />
        </div>
      </div>
      <div
        class="pf-c-toolbar__content pf-m-hidden"
        hidden=""
      >
        <div
          class="pf-c-toolbar__group"
        />
      </div>
    </div>
    <table
      aria-label="Simple"
      class="pf-c-table pf-m-grid-md pf-m-compact pf-m-sticky-header"
      data-ouia-component-id="OUIA-Generated-Table-1"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        />
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            colspan="1"
          >
            <div
              class="pf-l-bullseye"
            >
              <div
                class="pf-c-empty-state"
              >
                <div
                  class="pf-c-empty-state__content"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-c-empty-state__icon"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    />
                  </svg>
                  <h4
                    class="pf-c-title pf-m-lg"
                    data-ouia-component-id="OUIA-Generated-Title-2"
                    data-ouia-component-type="PF4/Title"
                    data-ouia-safe="true"
                  >
                    No results found
                  </h4>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;

exports[`single entry returned, both filters  1`] = `
<DocumentFragment>
  <section
    class="pf-c-page__main-section pf-m-light"
  >
    <div
      class="pf-l-level"
    >
      <div>
        <h1
          class="pf-c-title pf-m-2xl"
          data-ouia-component-id="OUIA-Generated-Title-3"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          Simple
        </h1>
      </div>
    </div>
  </section>
  <section
    class="pf-c-page__main-section"
  >
    <div
      class="pf-c-toolbar"
      data-ouia-component-id="OUIA-Generated-Toolbar-2"
      data-ouia-component-type="PF4/Toolbar"
      data-ouia-safe="true"
      id="pf-random-id-2"
    >
      <div
        class="pf-c-toolbar__content"
      >
        <div
          class="pf-c-toolbar__content-section"
        >
          <div
            class="pf-c-toolbar__group pf-m-toggle-group pf-m-show-on-xl"
          >
            <div
              class="pf-c-toolbar__toggle"
            >
              <button
                aria-controls="pf-random-id-2-expandable-content-2"
                aria-disabled="false"
                aria-haspopup="false"
                aria-label="Show Filters"
                class="pf-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-3"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="pf-c-toolbar__group pf-m-filter-group"
            >
              <div
                class="pf-c-toolbar__item"
              >
                <div
                  class="pf-c-input-group"
                >
                  <div
                    class="pf-c-search-input"
                  >
                    <div
                      class="pf-c-input-group"
                    >
                      <div
                        class="pf-c-search-input__bar"
                      >
                        <span
                          class="pf-c-search-input__text"
                        >
                          <span
                            class="pf-c-search-input__icon"
                          >
                            <svg
                              aria-hidden="true"
                              fill="currentColor"
                              height="1em"
                              role="img"
                              style="vertical-align: -0.125em;"
                              viewBox="0 0 512 512"
                              width="1em"
                            >
                              <path
                                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                              />
                            </svg>
                          </span>
                          <input
                            aria-label="Search input"
                            class="pf-c-search-input__text-input"
                            placeholder="Filter by name"
                            value=""
                          />
                        </span>
                      </div>
                      <button
                        aria-disabled="false"
                        aria-label="Search"
                        class="pf-c-button pf-m-control"
                        data-ouia-component-id="OUIA-Generated-Button-control-3"
                        data-ouia-component-type="PF4/Button"
                        data-ouia-safe="true"
                        type="submit"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 448 512"
                          width="1em"
                        >
                          <path
                            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="pf-c-toolbar__group pf-m-filter-group"
            >
              <div
                class="pf-c-toolbar__item"
              >
                <div
                  class="pf-c-select"
                  data-ouia-component-id="OUIA-Generated-Select-single-2"
                  data-ouia-component-type="PF4/Select"
                  data-ouia-safe="true"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-label="Options menu"
                    aria-labelledby=" pf-select-toggle-id-1"
                    class="pf-c-select__toggle"
                    id="pf-select-toggle-id-1"
                    type="button"
                  >
                    <div
                      class="pf-c-select__toggle-wrapper"
                    >
                      <span
                        class="pf-c-select__toggle-text"
                      >
                        Namespace
                      </span>
                    </div>
                    <span
                      class="pf-c-select__toggle-arrow"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="pf-c-toolbar__item"
              >
                <div
                  class="pf-c-input-group"
                >
                  <div
                    class="pf-c-search-input"
                  >
                    <div
                      class="pf-c-input-group"
                    >
                      <div
                        class="pf-c-search-input__bar"
                      >
                        <span
                          class="pf-c-search-input__text"
                        >
                          <span
                            class="pf-c-search-input__icon"
                          >
                            <svg
                              aria-hidden="true"
                              fill="currentColor"
                              height="1em"
                              role="img"
                              style="vertical-align: -0.125em;"
                              viewBox="0 0 512 512"
                              width="1em"
                            >
                              <path
                                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                              />
                            </svg>
                          </span>
                          <input
                            aria-label="Search input"
                            class="pf-c-search-input__text-input"
                            placeholder="Filter by namespace"
                            value=""
                          />
                        </span>
                      </div>
                      <button
                        aria-disabled="false"
                        aria-label="Search"
                        class="pf-c-button pf-m-control"
                        data-ouia-component-id="OUIA-Generated-Button-control-4"
                        data-ouia-component-type="PF4/Button"
                        data-ouia-safe="true"
                        type="submit"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 448 512"
                          width="1em"
                        >
                          <path
                            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="pf-c-toolbar__item"
            >
              <button
                aria-disabled="false"
                aria-label="Manage Columns"
                class="pf-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-4"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64V160h160v256zm224 0H288V160h160v256z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="pf-c-toolbar__expandable-content"
          id="pf-random-id-2-expandable-content-2"
        >
          <div
            class="pf-c-toolbar__group"
          />
        </div>
      </div>
      <div
        class="pf-c-toolbar__content pf-m-hidden"
        hidden=""
      >
        <div
          class="pf-c-toolbar__group"
        />
      </div>
    </div>
    <table
      aria-label="Simple"
      class="pf-c-table pf-m-grid-md pf-m-compact pf-m-sticky-header"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-3"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            class=""
            scope="col"
          >
            Name
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-4"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-label="Name"
          >
            foo_name
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
